File: cpp_source_code/206.反转链表.cpp
// @before-stub-for-debug-begin
#include <vector>
#include <string>

using namespace std;
// @before-stub-for-debug-end

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
/**
 * 思路：主要是从前向后遍历，操作next指针。
 * 但是链表的主要难点是改变next之后，原来的next节点就找不到了，所以
 * 需要提前存一下。
 * 前后两个节点一起向后移动。
 * 链表返回头指针不相当于返回一整个链表，leetcode是用头指针把链表遍历了一遍
 *
 */
class Solution
{
public:
    ListNode *reverseList(ListNode *head)
    {
        ListNode *pre = nullptr;
        ListNode *cur = head;

        while (cur != nullptr)
        {
            ListNode *tmp = cur->next;
            cur->next = pre;
            pre = cur;
            cur = tmp;
        }
        return pre;
        cout << pre << endl;
    }
};

int main()
{
    ListNode *node1 = new ListNode(1);
    ListNode *node2 = new ListNode(2);
    node1->next = node2;
    node2->next = nullptr;
    Solution::reverseList(node1);

    return 0;
}

File: py_source_code/206.反转链表.py
#
# @lc app=leetcode.cn id=206 lang=python3
#
# [206] 反转链表
#

# @lc code=start
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
# class Solution:
#     def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
#         #双指针法
#         pre = None
#         cur = head
#         while (cur):
#             temp = cur.next
#             cur.next = pre
#             pre = cur
#             cur = temp
#         return pre
class Solution:
    def reverse(self, pre:ListNode, cur:ListNode) -> ListNode:
        if (cur == None):
            return pre
        temp = cur.next
        cur.next = pre
        return self.reverse(cur, temp)
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        #递归实现
        return self.reverse(None, head)
# @lc code=end

